napari_aicsimageio/core.py

```diff
@@ -57,53 +139,29 @@ def reader_function(
 
     # Open file and get data
     img = AICSImage(path)
-    print(
-        f"AICSImageIO: Image contains {len(img.scenes)} scenes. "
-        f"napari-aicsimageio currently only supports loading first scene, "
-        f"will load scene: '{img.current_scene}'."
-    )
 
-    data = _get_full_image_data(img, in_memory=in_memory)
+    # Check for multiple scenes
+    if len(img.scenes) > 1:
+        print(
+            f"AICSImageIO: Image contains {len(img.scenes)} scenes. "
+            f"Supporting more than the first scene is experimental. "
+            f"Select a scene from the list widget. There may be dragons!"
+        )
+        # Launch the list widget
+        _get_scenes(img, in_memory=in_memory)
 
-    # Catch None data
-    if data is None:
-        return None
+        # Return an empty LayerData list; ImgLayers will be handled via the widget.
+        # HT Jonas Windhager
+        return [(None,)]
```

Comment:
Seems like Talleys comments may resolve this one but if not, what happens if you just return an empty list? `[]`

Reply:
Jumping in here (just got the email notification :innocent:). The `[(None,)]` sentinel is the correct way to go, see https://github.com/napari/napari/blob/24d01f5fee450c881c5f8fce4d17395f9af2328e/napari/plugins/hook_specifications.py#L115-L118. It was introduced with/after discussion in https://github.com/napari/napari/pull/2206.

Reply:
I'm still not clear if I should replace
```
LayerData = Union[Tuple[types.ArrayLike, Dict[str, Any], str]]
PathLike = Union[str, List[str]]
ReaderFunction = Callable[[PathLike], List[LayerData]]
```
with 
```
if TYPE_CHECKING:
    from napari.types import LayerData, ReaderFunction, PathLike
```
Or just add that? I got strange behavior when I replaced it, so I think im missing something. Docs suggest that the `if` shouldn't fire at runtime, which is bad? Should I just import the types instead?

Reply:
I believe you should replace them. What was the stranger behavior?

Reply:
I'll try again, but before it broke everything with
```
ValueError: No plugin found capable of reading '/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif'.
```
And after reverting everything stayed broken, even when using a different conda env. Ended up uninstalling/reinstalling the plugin in both my conda environments.

Reply:
Yup, broken again.
<details>

```
ValueError                                Traceback (most recent call last)
~/Dev/miniforge3/envs/napari-test-old/lib/python3.9/site-packages/napari/_qt/qt_viewer.py in dropEvent(self=<napari._qt.qt_viewer.QtViewer object>, event=<PyQt5.QtGui.QDropEvent object>)
    994                 filenames.append(url.toString())
    995 
--> 996         self.viewer.open(filenames, stack=bool(shift_down))
        self.viewer.open = <bound method ViewerModel.open of Viewer(axes=Axes(visible=False, labels=True, colored=True, dashed=False, arrows=True), camera=Camera(center=(0.0, 0.0, 0.0), zoom=1.0, angles=(0.0, 0.0, 90.0), perspective=0.0, interactive=True), cursor=Cursor(position=(1.0, 1.0), scaled=True, size=1, style=<CursorStyle.STANDARD: 'standard'>), dims=Dims(ndim=2, ndisplay=2, last_used=0, range=((0, 2, 1), (0, 2, 1)), current_step=(0, 0), order=(0, 1), axis_labels=('0', '1')), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[], scale_bar=ScaleBar(visible=False, colored=False, ticks=True, position=<Position.BOTTOM_RIGHT: 'bottom_right'>, font_size=10.0, unit=None), text_overlay=TextOverlay(visible=False, color=<class 'numpy.ndarray'> (4,) float64, font_size=10.0, position=<TextOverlayPosition.TOP_LEFT: 'top_left'>, text=''), help='', status='Ready', tooltip=Tooltip(visible=False, text=''), theme='light', title='napari', mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x140807b80>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, keymap={'Control-Shift-C': <bound method QtViewer.toggle_console_visibility of <napari._qt.qt_viewer.QtViewer object at 0x11d782040>>})>
        filenames = ['/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif']
        global stack = undefined
        global bool = undefined
        shift_down = <PyQt5.QtCore.Qt.KeyboardModifiers object at 0x1785e6d60>
    997 
    998     def closeEvent(self, event):

~/Dev/miniforge3/envs/napari-test-old/lib/python3.9/site-packages/napari/components/viewer_model.py in open(self=Viewer(axes=Axes(visible=False, labels=True, col...._qt.qt_viewer.QtViewer object at 0x11d782040>>}), path=['/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif'], stack=False, plugin=None, layer_type=None, **kwargs={})
    912         for _path in paths:
    913             added.extend(
--> 914                 self._add_layers_with_plugins(
        self._add_layers_with_plugins = <bound method ViewerModel._add_layers_with_plugins of Viewer(axes=Axes(visible=False, labels=True, colored=True, dashed=False, arrows=True), camera=Camera(center=(0.0, 0.0, 0.0), zoom=1.0, angles=(0.0, 0.0, 90.0), perspective=0.0, interactive=True), cursor=Cursor(position=(1.0, 1.0), scaled=True, size=1, style=<CursorStyle.STANDARD: 'standard'>), dims=Dims(ndim=2, ndisplay=2, last_used=0, range=((0, 2, 1), (0, 2, 1)), current_step=(0, 0), order=(0, 1), axis_labels=('0', '1')), grid=GridCanvas(stride=1, shape=(-1, -1), enabled=False), layers=[], scale_bar=ScaleBar(visible=False, colored=False, ticks=True, position=<Position.BOTTOM_RIGHT: 'bottom_right'>, font_size=10.0, unit=None), text_overlay=TextOverlay(visible=False, color=<class 'numpy.ndarray'> (4,) float64, font_size=10.0, position=<TextOverlayPosition.TOP_LEFT: 'top_left'>, text=''), help='', status='Ready', tooltip=Tooltip(visible=False, text=''), theme='light', title='napari', mouse_move_callbacks=[], mouse_drag_callbacks=[], mouse_double_click_callbacks=[], mouse_wheel_callbacks=[<function dims_scroll at 0x140807b80>], _persisted_mouse_event={}, _mouse_drag_gen={}, _mouse_wheel_gen={}, keymap={'Control-Shift-C': <bound method QtViewer.toggle_console_visibility of <napari._qt.qt_viewer.QtViewer object at 0x11d782040>>})>
        _path = '/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif'
        kwargs = {}
        plugin = None
        layer_type = None
    915                     _path, kwargs, plugin=plugin, layer_type=layer_type
    916                 )

~/Dev/miniforge3/envs/napari-test-old/lib/python3.9/site-packages/napari/components/viewer_model.py in _add_layers_with_plugins(self=Viewer(axes=Axes(visible=False, labels=True, col...._qt.qt_viewer.QtViewer object at 0x11d782040>>}), path_or_paths='/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif', kwargs={}, plugin=None, layer_type=None)
    957         from ..plugins.io import read_data_with_plugins
    958 
--> 959         layer_data, hookimpl = read_data_with_plugins(
        layer_data = undefined
        hookimpl = undefined
        read_data_with_plugins = <function read_data_with_plugins at 0x17aba1160>
        path_or_paths = '/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif'
        plugin = None
    960             path_or_paths, plugin=plugin
    961         )

~/Dev/miniforge3/envs/napari-test-old/lib/python3.9/site-packages/napari/plugins/io.py in read_data_with_plugins(path='/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif', plugin=None)
    134 
    135         # TODO: change to a warning notification in a later PR
--> 136         raise ValueError(message)
        global ValueError = undefined
        message = "No plugin found capable of reading '/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif'."
    137 
    138     if errors:

ValueError: No plugin found capable of reading '/Users/piotrsobolewski/Documents/Leica DMi8/2021040_H33342_L929_24w_A3_tilescan.lif'.

```
</details>

Reply:
Huh..... Can you push your most up-to-date changes. I can take a look at this tomorrow if not solved

Reply:
When you open napari without loading a file does `napari-aicsimageio` show up in the plugin list?

Reply:
sorry, this napari's bad error message problem.  I'll explain the typing issue in just minute, but the reason that this appears to be "permanently breaking" the plugin is because:
a) as is, your plugin is not importable (reason why in just a second)
b) when napari finds a plugin that can't even be imported, it adds it to the "blocked" list.
c) that used to make a lot of sense, but then later we added "persistence" to our settings... so now it's blocked forever :joy: unless you use `napari --reset` (or, if you have some settings you'd rather not lose, go into your settings and manually remove this from the blocked plugins).  This is totally our fault and I'll submit a fix soon.

the _reason_ it's not importing (sorry I didn't explain this before, it's a good general lesson):
While it's generally good practice to use `if TYPE_CHECKING:`... there's one more thing you need to be aware of.  At runtime `if TYPE_CHECKING:` is `False` and so none of those variables get declared.  so the actual traceback if you try to import your plugin (directly, no via napari, which is a good thing to do when debugging):

```pytb
aicsimageio-in-memory
  error: PluginImportError('Error while importing module napari_aicsimageio.in_memory')
  cause: NameError("name 'PathLike' is not defined")
  traceback:
     File "/Users/talley/miniconda3/envs/napdev/lib/python3.9/site-packages/napari_plugin_engine/manager.py", line 292, in _load_and_register
       module = load(mod_name)
     File "/Users/talley/miniconda3/envs/napdev/lib/python3.9/site-packages/napari_plugin_engine/manager.py", line 1005, in load
       module = importlib.import_module(match.group('module'))
     File "/Users/talley/miniconda3/envs/napdev/lib/python3.9/importlib/__init__.py", line 127, in import_module
       return _bootstrap._gcd_import(name[level:], package, level)
     File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
     File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
     File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
     File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
     File "<frozen importlib._bootstrap_external>", line 850, in exec_module
     File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
     File "/Users/talley/Dropbox (HMS)/Python/forks/napari-aicsimageio/napari_aicsimageio/in_memory.py", line 8, in <module>
       from . import core
     File "/Users/talley/Dropbox (HMS)/Python/forks/napari-aicsimageio/napari_aicsimageio/core.py", line 127, in <module>
       path: PathLike, in_memory: bool, scene_name: Optional[str] = None
```

so basically, later down in the file, python is trying to evaluate the `PathLike` type hint ... but you really don't need it to.  So you need to make it a "[forward reference](https://www.python.org/dev/peps/pep-0484/#forward-references)" (fancy word for just make it a string), in one of two ways:

**either of this will work**

option 1:  (use `from __future__ import annotations`) 

```py
from __future__ import annotations

if TYPE_CHECKING:
    from napari.types import LayerData, ReaderFunction, PathLike

def get_reader(path: PathLike, in_memory: bool) -> Optional[ReaderFunction]: ...
```

option 2: (use strings)

```py
if TYPE_CHECKING:
    from napari.types import LayerData, ReaderFunction, PathLike

def get_reader(path: 'PathLike', in_memory: bool) -> Optional['ReaderFunction']: ...
```

Reply:
Wowza! Talk about a can of worms!
@JacksonMaxfield which way would you like me to proceed?

Reply:
I'd recommend option 2 for now (unless you have a codebase with a lot of `from __future__ import annotations` it can be a little mysterious.  just make your annotations strings

Reply:
I tried option 2: I quoted every occurrence of those 3 types that are in the `if`.
Still getting the same issue though.
At least it's an easy fix thanks to your explanation above.
Option 3 does work though...

Reply:
I vote option 3 then. Whatever works, works!

Reply:
Done.
(I swear I've now spent more time on the linting than actually getting the feature implemented 🤣 )

Reply:
Sounds about right. Getting typed Python stuff correct can be especially hard.

Reply:
@psobolewskiPhD I think that line which case that try 1 and 2 fail for you is here https://github.com/AllenCellModeling/napari-aicsimageio/blob/dacc2727b5b897c53705589b2fef00dce4f908fc/napari_aicsimageio/in_memory.py#L14

I suggest fixing `in_memory.py` and `out_of_memory.py` by change import `from . import core` to `from napari.types import PathLike` and then fix the signature of `napari_hook_implementation`.

Reply:
What do you think @JacksonMaxfield ? This seems like a good idea, to get the types from napari.types.

Reply:
More explanarion. During runtime, when `TYPE_CHECKING` is `False` then `PathLike` and `ReaderFunction` are undefined in `core` module. 
Another option to solve this is to change to 

```  
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import annotation

from typing import Optional

from napari_plugin_engine import napari_hook_implementation

if TYPE_CHECKING:
    from . import core

###############################################################################


@napari_hook_implementation
def napari_get_reader(path: core.PathLike) -> Optional[core.ReaderFunction]:
    return core.get_reader(path, in_memory=True)
```